consensus/src/chained_bft/persistent_storage.rs

```diff
@@ -186,23 +186,21 @@ impl<T: Payload> RecoveryData<T> {
         // sort by round to guarantee the topological order of parent <- child
         blocks.sort_by_key(Block::round);
         let root_from_consensus = {
-            let id_to_round: HashMap<_, _> = blocks
+            let id_to_id_and_round: HashMap<_, _> = blocks
                 .iter()
-                .map(|block| (block.id(), block.round()))
+                .map(|block| (block.id(), (block.id(), block.round())))
                 .collect();
-            let mut round_and_id = None;
-            for qc in quorum_certs.iter() {
-                if let Some(committed_block_id) = qc.committed_block_id() {
-                    if let Some(round) = id_to_round.get(&committed_block_id) {
-                        match round_and_id {
-                            Some((r, _)) if r > round => (),
-                            _ => round_and_id = Some((round, committed_block_id)),
-                        }
-                    }
-                }
-            }
-            match round_and_id {
-                Some((_, id)) => id,
+
+            let root_id_and_round = quorum_certs
+                .iter()
+                .flat_map(|qc| {
```

Comment:
Just wondering: why is it a `flat_map` and not a `map`? Aren't you returning a single value here?

Reply:
`flat_map` will [unwrap the options](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=02c8ae59a99e81938f7b58e604bd7542) for me here.

Reply:
TIL, thanks!